File: app/Chatbridge/Core/Profiles.swift
import CryptoKit
import Foundation

/// A declarative description of how to read one app surface. Profiles ship
/// inside the signed app bundle, so they can't be swapped without breaking the
/// code signature; the checksum ties every stored record to the exact revision.
nonisolated struct Profile: @unchecked Sendable {
    let name: String
    let version: Int
    let data: [String: Any]
    let checksum: String

    var key: String { "\(name)@\(version)" }
    var appID: String { (data["app"] as? [String: Any])?["id"] as? String ?? name }
    var schema: String { data["schema"] as? String ?? "" }
    var hosts: [String] { (data["match"] as? [String: Any])?["hosts"] as? [String] ?? [] }
    var canary: [String: Any] { data["canary"] as? [String: Any] ?? [:] }
    var isDevelopmentFixture: Bool { appID.hasPrefix("dev.chatbridge.") }
}

nonisolated struct ProfileError: Error, CustomStringConvertible { let description: String }

nonisolated final class ProfileLibrary: @unchecked Sendable {
    static let shared = ProfileLibrary()

    let schemas: [String: [String: Any]]
    let profiles: [String: Profile]

    private init() {
        var schemas: [String: [String: Any]] = [:]
        for url in Bundle.main.urls(forResourcesWithExtension: "json", subdirectory: "schemas") ?? [] {
            if let doc = (try? JSONSerialization.jsonObject(with: Data(contentsOf: url))) as? [String: Any], let name = doc["schema"] as? String {
                schemas[name] = doc
            }
        }
        var profiles: [String: Profile] = [:]
        for url in Bundle.main.urls(forResourcesWithExtension: "json", subdirectory: "profiles") ?? [] {
            guard let raw = try? Data(contentsOf: url),
                  let doc = (try? JSONSerialization.jsonObject(with: raw)) as? [String: Any] else { continue }
            do {
                try Self.validate(doc, schemas: schemas)
                let profile = Profile(
                    name: doc["profile"] as? String ?? "", version: doc["profileVersion"] as? Int ?? 0, data: doc,
                    checksum: SHA256.hash(data: raw).map { String(format: "%02x", $0) }.joined())
                profiles[profile.key] = profile
            } catch {
                DebugLog.write("profiles: rejected \(url.lastPathComponent): \(error)")
            }
        }
        self.schemas = schemas
        self.profiles = profiles
        DebugLog.write("profiles: loaded \(profiles.keys.sorted().joined(separator: ", "))")
    }

    static func validate(_ data: [String: Any], schemas: [String: [String: Any]]) throws {
        for key in ["profile", "profileVersion", "app", "match", "schema", "canary"] where data[key] == nil {
            throw ProfileError(description: "missing key \(key)")
        }
        guard let schemaName = data["schema"] as? String, let schema = schemas[schemaName],
              let fields = schema["fields"] as? [String: [String: Any]] else { throw ProfileError(description: "unknown schema") }
        let method = data["method"] as? String ?? "dom"
        var produced: Set<String>
        switch method {
        case "dom":
            guard let extract = data["extract"] as? [String: Any], extract["item"] != nil,
                  let f = extract["fields"] as? [String: Any] else { throw ProfileError(description: "extract needs item and fields") }
            produced = Set(f.keys).union((extract["context"] as? [String: Any] ?? [:]).keys)
        case "page-store":
            guard let store = data["store"] as? [String: Any], let f = store["fields"] as? [String: Any],
                  (store["root"] as? String == "window" ? store["path"] : store["collection"]) != nil
            else { throw ProfileError(description: "store needs a source (collection, or window path) and fields") }
            produced = Set(f.keys)
        default:
            throw ProfileError(description: "unknown method \(method)")
        }
        let unknown = produced.subtracting(fields.keys)
        guard unknown.isEmpty else { throw ProfileError(description: "fields not in schema: \(unknown.sorted())") }
        let identity = fields.filter { $0.value["identity"] as? Bool == true }.map(\.key)
        guard Set(identity).isSubset(of: produced) else { throw ProfileError(description: "identity fields not extracted") }
        let checked = Set(((data["canary"] as? [String: Any])?["fillRate"] as? [String: Any] ?? [:]).keys)
        guard checked.isSubset(of: produced) else { throw ProfileError(description: "canary checks fields that are never extracted") }
    }

    /// Profiles a page on `host` may use. Development fixtures are never offered to real hosts.
    func profiles(forHost host: String, isQuarantined: (String) -> Bool) -> [[String: Any]] {
        profiles.values.filter { $0.hosts.contains(host) && !isQuarantined($0.key) }.map(\.data)
    }
}

File: dev/harness/host/bridge/profiles.py
"""Load and validate app profiles and the normalized schemas they target."""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[4]
PROFILES_DIR = REPO_ROOT / "profiles"
SCHEMAS_DIR = REPO_ROOT / "schemas"

REQUIRED_KEYS = ("profile", "profileVersion", "app", "match", "schema", "canary")
METHODS = ("dom", "page-store")


class ProfileError(ValueError):
    pass


@dataclass(frozen=True)
class Profile:
    name: str
    version: int
    data: dict
    checksum: str  # sha256 of the file bytes: pins stored data to an exact profile revision

    @property
    def key(self) -> str:
        # Several versions of one profile can coexist, each matching a
        # different range of upstream app versions.
        return f"{self.name}@{self.version}"

    @property
    def app_id(self) -> str:
        return self.data["app"]["id"]

    @property
    def schema(self) -> str:
        return self.data["schema"]

    @property
    def hosts(self) -> list[str]:
        return self.data["match"].get("hosts", [])


def load_schemas(directory: Path = SCHEMAS_DIR) -> dict[str, dict]:
    schemas = {}
    for path in sorted(directory.glob("*.json")):
        doc = json.loads(path.read_text())
        schemas[doc["schema"]] = doc
    return schemas


def validate(data: dict, schemas: dict[str, dict]) -> None:
    missing = [k for k in REQUIRED_KEYS if k not in data]
    if missing:
        raise ProfileError(f"missing keys: {missing}")
    if data["schema"] not in schemas:
        raise ProfileError(f"unknown schema {data['schema']!r}")
    allowed = set(schemas[data["schema"]]["fields"])
    method = data.get("method", "dom")
    if method not in METHODS:
        raise ProfileError(f"unknown method {method!r}")
    if method == "dom":
        extract = data.get("extract", {})
        if "item" not in extract or "fields" not in extract:
            raise ProfileError("extract needs 'item' and 'fields'")
        produced = set(extract["fields"]) | set(extract.get("context", {}))
    else:
        store = data.get("store", {})
        source = ("path",) if store.get("root") == "window" else ("require", "collection")
        if missing := [k for k in (*source, "fields") if k not in store]:
            raise ProfileError(f"store needs {missing}")
        produced = set(store["fields"])
    if unknown := produced - allowed:
        raise ProfileError(f"fields not in schema {data['schema']}: {sorted(unknown)}")
    identity = [n for n, f in schemas[data["schema"]]["fields"].items() if f.get("identity")]
    if not set(identity) <= produced:
        raise ProfileError(f"identity fields not extracted: {identity}")
    if unknown := set(data["canary"].get("fillRate", {})) - produced:
        raise ProfileError(f"canary checks fields that are never extracted: {sorted(unknown)}")


def load_profiles(directory: Path = PROFILES_DIR, schemas: dict[str, dict] | None = None) -> dict[str, Profile]:
    schemas = schemas if schemas is not None else load_schemas()
    profiles: dict[str, Profile] = {}
    for path in sorted(directory.glob("*.json")):
        raw = path.read_bytes()
        data = json.loads(raw)
        try:
            validate(data, schemas)
        except ProfileError as e:
            raise ProfileError(f"{path.name}: {e}") from e
        profile = Profile(data["profile"], int(data["profileVersion"]), data, hashlib.sha256(raw).hexdigest())
        if profile.key in profiles:
            raise ProfileError(f"{path.name}: duplicate profile {profile.key!r}")
        profiles[profile.key] = profile
    return profiles
